chpl/chpl-service/src/main/java/gov/healthit/chpl/upload/listing/ListingUploadManager.java

```diff
@@ -0,0 +1,377 @@
+    public List<ListingUpload> parseUploadFile(MultipartFile file) throws ValidationException {
+        List<CSVRecord> allCsvRecords = getFileAsCsvRecords(file);
+        int headingRowIndex = uploadUtil.getHeadingRecordIndex(allCsvRecords);
+        if (headingRowIndex < 0) {
+            throw new ValidationException(msgUtil.getMessage("listing.upload.noHeadingFound"));
+        }
+        CSVRecord headingRecord = uploadUtil.getHeadingRecord(allCsvRecords);
+        List<CSVRecord> allListingRecords = allCsvRecords.subList(headingRowIndex + 1, allCsvRecords.size());
+        if (headingRecord == null) {
+            LOGGER.warn("Cannot continue parsing upload file " + file.getName() + " without heading.");
+            throw new ValidationException(msgUtil.getMessage("listing.upload.noHeadingFound"));
+        }
+        checkRequiredHeadings(headingRecord);
+
+        List<ListingUpload> uploadedListings = new ArrayList<ListingUpload>();
+        List<List<CSVRecord>> groupedListingRecords = new ArrayList<List<CSVRecord>>();
+        int fileStartIndex = 0;
+        while (fileStartIndex < allListingRecords.size()) {
+            List<CSVRecord> currListingRecords = getNextListingRecordGroup(fileStartIndex, headingRecord, allListingRecords);
+            groupedListingRecords.add(currListingRecords);
+            fileStartIndex += currListingRecords.size();
+        }
+        for (List<CSVRecord> listingRecords : groupedListingRecords) {
+            checkRequiredFields(headingRecord, listingRecords);
+        }
+        groupedListingRecords.stream()
+                .map(listingRecords -> createListingUploadMetadata(headingRecord, listingRecords))
+                .forEach(listingUploadMetadata -> {
+                    uploadedListings.add(listingUploadMetadata);
+                });
+        return uploadedListings;
+    }
+
+    @Transactional
+    @PreAuthorize("@permissions.hasAccess(T(gov.healthit.chpl.permissions.Permissions).LISTING_UPLOAD, "
+            + "T(gov.healthit.chpl.permissions.domains.ListingUploadDomainPerissions).CREATE, #uploadMetadata)")
+    public ListingUpload createOrReplaceListingUpload(ListingUpload uploadMetadata) throws ValidationException,
+        JsonProcessingException, EntityRetrievalException, EntityCreationException {
+        if (StringUtils.isEmpty(uploadMetadata.getChplProductNumber())) {
+            throw new ValidationException(msgUtil.getMessage("listing.upload.missingChplProductNumber"));
+        } else if (uploadMetadata.getAcb() == null || uploadMetadata.getAcb().getId() == null) {
+            throw new ValidationException(msgUtil.getMessage("listing.upload.missingAcb"));
+        }
+
+        //Should we actually only allow unique CHPL IDs here?
+        //It would keep functionality the same, but a file with new developer codes can have multiple
+        //listings with the same pending CHPL ID and the latest one would overwrite the earlier one.
```

Comment:
Should this question be put in business terms and presented to PO?

Reply:
Added a checkbox item